### DISCUSSION TITLE
Keymap file failed to load

### DISCUSSION THREAD

**Discussion by muralikodali** [qutebrowser/qutebrowser#7896] [Q&A]
qutebrowser version is 3.0.0-1

My OS is Arch Linux.

After updating to version 3.0.0 , qutebrowser is failing to start some times.

```qutebrowser --temp-basedir``` is giving following error.

```
xkbcommon: ERROR: Couldn't find file "symbols/te" in include paths
xkbcommon: ERROR: 1 include paths searched:
xkbcommon: ERROR:       /usr/share/X11/xkb
xkbcommon: ERROR: Abandoning symbols file "(unnamed)"
xkbcommon: ERROR: Failed to compile xkb_symbols
xkbcommon: ERROR: Failed to compile keymap
[20572:20592:0831/115913.323801:FATAL:xkb_keyboard_layout_engine.cc(652)] Keymap file failed to load: in,te-,eng
fish: Job 1, 'qutebrowser --temp-basedir' terminated by signal SIGTRAP (Trace or breakpoint trap)
```
this problem is not happening every time. it is failing to start some times and loading as expected some times.

**Comment by The-Compiler** [comment]
That's weird. Do you have an idea what that `in,te-,eng` keyboard layout is exactly, or what the `symbols/te` is that it can't find? I know nothing about keyboard layouts to be honest.

Also, how did you reproduce this with `--temp-basedir` then? Did you just have to try a couple of times? Or are there scenarios where it's reproducible even multiple times in a row, and scenarios where it's not?

**Comment by muralikodali** [comment]
eng: US english
in: indian english
te-: telugu, a regional language in india

launching qutebrowser from the terminal, giving the output showing "Keymap file failed to load".

launching from gui, it leaves no clue for failing.

In a session, it is happening in a row. If i shut my computer and launch again, qutebrowser may launch as expected.

**Comment by bitraid** [comment]
According to `man 7 xkeyboard-config`, there in no `eng` layout (it's `us` or a variant like `us(int)`), the Indian English is `in(eng)` (whereas the description for `in` is Indian), and the telugu layout is `in(tel)` (or `in(tel-kagapa)`, `in(tel-sarala)`). Do you have files named `eng` or `te` in `/usr/share/X11/xkb/symbols/`?

**Comment by muralikodali** [comment]
now, after a fresh system reboot, qutebrowser is opening as intended.
i have checked /usr/share/X11/xkb/symbols. no files named eng or te are found. in and us files are found.

**Comment by bitraid** [comment]
What is your desktop environment? How do you set the keyboard layout? Can you show the output of `setxkbmap -print -verbose 10`?

**Comment by muralikodali** [comment]
MY desktop environment is i3.

my i3 keyboard configurtion looks as below :  

```
# keyboard layout toggle
set $keyboard_layout Keyboard Layout: (F1) in (F2) tel (F3) us
mode "$keyboard_layout" {
bindsym F1 exec --no-startup-id setxkbmap in -variant eng ,mode "default"
bindsym F2 exec --no-startup-id setxkbmap tel ,mode "default"
bindsym F3 exec --no-startup-id setxkbmap us ,mode "default"
# back to normal: Enter or Escape
bindsym Return mode "default"
bindsym Escape mode "default"
}

# toggle keyboard layout
bindsym $mod+Tab mode "$keyboard_layout"

```

output of ```setxkbmap -print -verbose 10``` is as below : 

```
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc104
layout:     us
options:    compose:ralt,terminate:ctrl_alt_bksp,compose:rmet
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+inet(evdev)+compose(ralt)+terminate(ctrl_alt_bksp)
geometry:   pc(pc104)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+us+inet(evdev)+compose(ralt)+terminate(ctrl_alt_bksp)"	};
	xkb_geometry  { include "pc(pc104)"	};
};

```

**Comment by bitraid** [comment]
The most logical explanation would be that, at some point, something (a script maybe?) messes up your keyboard layout (BTW, do you know what `,mode "default"` does? I could not find any documentation for it). Have you noticed running something else before qutebrowser starts failing? You should also try running `setxkbmap -print -verbose 10` next time you get the error. Can you reproduce the issue by running `setxkbmap ,in,te-,eng; qutebrowser --temp-basedir`? Also could you check if the file `/etc/X11/xorg.conf.d/00-keyboard.conf` exists, and if so, what are its contents?

**Comment by bitraid** [comment]
I managed to reproduce the error by setting an invalid layout:
```
xkbcommon: ERROR: Couldn't find file "symbols/eng" in include paths
xkbcommon: ERROR: 1 include paths searched:
xkbcommon: ERROR: 	/usr/share/X11/xkb
xkbcommon: ERROR: Abandoning symbols file "(unnamed)"
xkbcommon: ERROR: Failed to compile xkb_symbols
xkbcommon: ERROR: Failed to compile keymap
[13161:13191:0902/074255.407792:FATAL:xkb_keyboard_layout_engine.cc(652)] Keymap file failed to load: ,eng
Trace/breakpoint trap (core dumped)
```
So, this is most likely not an issue with qutebrowser.

**Comment by muralikodali** [comment]
running the command ```setxkbmap ,in,te-,eng; qutebrowser --temp-basedir``` results in failure of starting qutebrowser.

contents of /etc/X11/xorg.conf.d/00-keyboard.conf are shown below : 
```
Section "InputClass"
        Identifier "system-keyboard"
        MatchIsKeyboard "on"
        Option "XkbLayout" "in,te"
        Option "XkbModel" "pc104"
	Option "XkbVariant" ",eng"
        Option "XkbOptions" "terminate:ctrl_alt_bksp,compose:rmet"
EndSection
```
this kind of problem is happening only after updating qutebrowser to 3.0 version.

**Comment by The-Compiler** [comment]
Can you show the output of `setxkbmap -print -verbose 10` when it happens vs. when it does not?

I agree this is most likely an issue with your setup and not qutebrowser. Perhaps the older Chromium version handled things in a different way, but the actual culprit is probably not Chromium (and certainly not qutebrowser, given that the error happens in Chromium).

**Comment by The-Compiler** [comment]
I found a lot of related stuff with a cursory search:

- [[QTBUG-101201] Failed to compile xkb_symbols with multiple layout - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-101201)
- [[QTBUG-105124] qtwebengine crashes when `/usr/share/X11/xkb` is empty - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-105124)
- https://github.com/NixOS/nixpkgs/pull/227077
- https://github.com/MirServer/graphics-core22/issues/7
- https://github.com/flathub/com.viber.Viber/issues/60
- https://github.com/ankitects/anki/issues/1814
- https://github.com/textbrowser/dooble/issues/101

@muralikodali could you please also show your full `:version` output?

**Comment by muralikodali** [comment]
version info url: https://bpa.st/LXHBS

following is the output of ```setxkbmap -print -verbose 10``` : 

on successful start of qutebrowser : 

```
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc104
layout:     us
options:    compose:ralt,terminate:ctrl_alt_bksp,compose:rmet
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+inet(evdev)+compose(ralt)+terminate(ctrl_alt_bksp)
geometry:   pc(pc104)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+us+inet(evdev)+compose(ralt)+terminate(ctrl_alt_bksp)"	};
	xkb_geometry  { include "pc(pc104)"	};
};
```

on failure to start the browser : 

```
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc104
layout:     ,in,te-,eng
options:    compose:ralt,terminate:ctrl_alt_bksp,compose:rmet
Trying to build keymap using the following components:
keycodes:   evdev
types:      complete
geometry:   pc(pc104)
xkb_keymap {
	xkb_keycodes  { include "evdev"	};
	xkb_types     { include "complete"	};
	xkb_geometry  { include "pc(pc104)"	};
};

```

**Comment by bitraid** [comment]
The reason that you get the error only with the new version of qutebrowser, is because of Qt6 - I just tried Qt5 with an invalid layout, and there is no error. Your `00-keyboard.conf` doesn't look right to me, but since you are calling directly `setxkbmap`, it gets ignored (after the first call). You can see that by running `setxkbmap us -verbose`:
```
Warning! Multiple definitions of keyboard layout
         Using command line, ignoring X server
...
```
Running the above command (or any command that sets a valid layout), should also make qutebrowser run without error again. I would suggest to replace the contents of `00-keyboard.conf` with:
```
Section "InputClass"
        Identifier "system-keyboard"
        MatchIsKeyboard "on"
        Option "XkbLayout" "us"
        Option "XkbModel" "pc104"
        Option "XkbVariant" ""
        Option "XkbOptions" "terminate:ctrl_alt_bksp,compose:rmet"
EndSection
```

**Comment by The-Compiler** [comment]
Moving this to a discussion since it's not a qutebrowser bug.

**Comment by bitraid** [comment]
To demonstrate that qutebrowser works correctly with multiple layouts/variants when they are configured correctly, run:
`setxkbmap -layout "us,in,in" -variant ",eng,tel" -option "grp:alt_shift_toggle"; qutebrowser`
(use `Alt-Shift` to switch between English, Indian/English, Indian/Telugu).
The corresponding `00-keyboard.conf`, is:
```
Section "InputClass"
        Identifier "system-keyboard"
        MatchIsKeyboard "on"
        Option "XkbLayout" "us,in,in"
        Option "XkbModel" "pc104"
        Option "XkbVariant" ",eng,tel"
        Option "XkbOptions" "grp:alt_shift_toggle"
EndSection
```

**Comment by muralikodali** [comment]
@bitraid thanks for your suggestion.

changed 00-keyboard.conf as suggested by you. 

now qutebrowser is starting without errors.
